Add too many features!

mtalaeii opened the pull request:
- Add CallbackQuery class for handle callbackQueries updates with 2 bound method(edit and answer)
- Add support for CallbackQueries in UpdateHandler class
- Add FilterCallback & FilterCallbackRegex class for handle callbackQueries texts
- Change structure for FilterFromAdmin to handle also callbackQueries requests
- add 2 method edit and translate to Message class

mtalaeii committed:
Add to many features!
- Add CallbackQuery class for handle callbackQueries updates with 2 bound method(edit and answer)
- Add support for CallbackQueries in UpdateHandler class
- Add FilterCallback & FilterCallbackRegex class for handle callbackQueries texts
- Change structure for FilterFromAdmin to handle also callbackQueries requests
- add 2 method edit and translate to Message class

mtalaeii committed:
Merge branch 'danog:v8' into v8

error1001es commented:
Typed property danog\MadelineProto\Ipc\Server::$callback must not be accessed before initialization

danog commented on src/EventHandler/Message.php:
```diff
@@ -259,6 +261,78 @@ public function delReaction(int|string $reaction): array
+    public function translate(
+        string $toLang = 'en'
+    ): string {
+        Assert::notEmpty($this->message);
```
Just return the message itself if it's empty

error1001es replied:
error while i using `$MadelineProto = new API('session.madeline');`

mtalaeii replied:
its ok i've also forgot to write phpdoc for CallbackQuery class methods ... will fix

mtalaeii replied:
@error1001es  please open an issue with full details (your code and MadelineProto.log file) 
this is not a good place to say your issue!!

mtalaeii committed:
Fix the translate method bug
- Add phpDoc for CallbackQuery class methods

danog commented on src/EventHandler/CallbackQuery.php:
```diff
@@ -0,0 +1,122 @@
+        parent::__construct($API);
+        $this->queryId = $rawCallback['query_id'];
+        $this->userId = $rawCallback['user_id'];
+        $this->chatId = $rawCallback['peer']['user_id'] ?? $rawCallback['peer']['chat_id'];
```
Please use `$API->getIdInternal($rawCallback['peer']);`

danog commented on src/EventHandler/Message.php:
```diff
@@ -259,6 +261,80 @@ public function delReaction(int|string $reaction): array
+     *
+     */
+    public function translate(
+        string $toLang = 'en'
```
It probably makes sense to have no default here

danog commented on src/MTProtoTools/UpdateHandler.php:
```diff
@@ -334,6 +335,7 @@ public function wrapUpdate(array $update): ?Update
     {
         return match ($update['_']) {
             'updateNewChannelMessage', 'updateNewMessage', 'updateNewScheduledMessage', 'updateEditMessage', 'updateEditChannelMessage' => $this->wrapMessage($update['message']),
+            'updateBotCallbackQuery' => new CallbackQuery($this, $update),
```
You should also handle updateInlineBotCallbackQuery

mtalaeii replied:
ahh its little different because it not contain a chatId here but ok i will add it too!

danog commented on src/EventHandler/CallbackQuery.php:
```diff
@@ -0,0 +1,122 @@
+     */
+    public readonly int $chatInstance;
+    /** @var string Callback data */
+    public readonly string $data;
```
Should be ?string

danog commented on src/EventHandler/CallbackQuery.php:
```diff
@@ -0,0 +1,122 @@
+use danog\MadelineProto\MTProto;
+use danog\MadelineProto\ParseMode;
+
+final class CallbackQuery extends Update
```
IMO It's a good idea to rename this to AbstractButtonQuery, and create`AbstractChatButtonQuery` and `AbstractInlineButtonQuery` classes, mostly empty except for a `$chatId`+`$messageId` in AbstractChatButtonQuery and a private $messageId in `AbstractInlineButtonQuery` (we shouldn't expose it, it only contains internal info used by the answer method).

Then four final classes specialize the Abstract(Inline)ButtonQuery to `(Inline)(Game|Callback)Query`, where the game and callback have either a `$data` or a `$gameShortName` property.

danog commented on src/EventHandler/CallbackQuery.php:
```diff
@@ -0,0 +1,122 @@
+                'no_webpage' => $noWebpage
+            ]
+        );
+        if (isset($result['_'])) {
```
This conditional is useless, editMessage always returns a single constructor (unlike sendMessage which may make multiple method calls).

mtalaeii replied:
you mean i must remove this condition and return results directly?

danog replied:
You must keep only the body of the condition, remove everything after it

danog commented on src/EventHandler/Filter/FilterCallbackRegex.php:
```diff
@@ -0,0 +1,27 @@
+use Webmozart\Assert\Assert;
+
+#[Attribute(Attribute::TARGET_METHOD)]
+class FilterCallbackRegex extends Filter
```
LGTM, but I'm not too convinced this is needed :)

mtalaeii replied:
you mean i remove entire FilterCallbackRegex in this case?! why?
we have talk about this in telegram and i have example use of this pattern its not useless

danog commented on src/EventHandler/Message.php:
```diff
@@ -259,6 +261,80 @@ public function delReaction(int|string $reaction): array
+     *
+     * @param string $message New message
+     * @param array|null $replyMarkup Reply markup for inline keyboards
+     * @param array|null $entities Message entities for styled text
```
Please don't include this

danog commented on src/EventHandler/Message.php:
```diff
@@ -259,6 +261,80 @@ public function delReaction(int|string $reaction): array
+                'no_webpage' => $noWebpage
+            ]
+        );
+        if (isset($result['_'])) {
```
As before, this is always true

danog reviewed (changes requested):
LGTM, except for a few small comments :)

mtalaeii committed:
Do all tasks
- separate ChatButtonQuery ChatGameQuery InlineButtonQuery and InlineGameQuery from CallbackQuery class
- add private wrapQuery to handle both inline and chat queries
- some changes to some methods and some fixes ...

mtalaeii reviewed (commented):
.

mtalaeii committed:
Update InlineGameQuery.php

mtalaeii committed:
Fix some bug

mtalaeii committed:
Fix
- Forgot to add constructor for ChatGameQuery

danog commented on src/EventHandler/Message.php:
```diff
@@ -289,17 +289,13 @@ public function translate(
      */
     public function edit(
         string    $message,
-        ?array    $replyMarkup = null,
```
replyMarkup is fine

danog commented on src/MTProtoTools/UpdateHandler.php:
```diff
@@ -335,10 +339,30 @@ public function wrapUpdate(array $update): ?Update
+    private function wrapQuery(array $callback): ?AbstractQuery
+    {
+        if ($callback['_'] == 'updateBotCallbackQuery') {
+            if (!isset($callback['game_short_name']) && !empty($callback['game_short_name'])) {
```
if (isset(...))

danog commented on src/MTProtoTools/UpdateHandler.php:
```diff
@@ -335,10 +339,30 @@ public function wrapUpdate(array $update): ?Update
+            return new ChatButtonQuery($this, $callback);
+        }
+        if ($callback['_'] == 'updateInlineBotCallbackQuery') {
+            if (!isset($callback['game_short_name']) && !empty($callback['game_short_name'])) {
```
if (isset(...))

danog commented on src/EventHandler/AbstractGameQuery.php:
```diff
@@ -0,0 +1,16 @@
+
+    public function __construct(MTProto $API, array $rawCallback)
+    {
+        parent::__construct($API);
```
$API, $rawCallback

danog commented on src/EventHandler/AbstractQuery.php:
```diff
@@ -0,0 +1,32 @@
+    /** @internal */
+    public function __construct(MTProto $API, array $rawCallback)
+    {
+        parent::__construct($API);
```
$API, $rawCallback

mtalaeii committed:
Some fixes
- Also remove empty assertion from edit bound method as it doesn't need here

mtalaeii reviewed (commented):
Fixed!

danog commented on src/EventHandler/ChatButtonQuery.php:
```diff
@@ -0,0 +1,82 @@
+    /** @var int Message ID */
+    public readonly int $messageId;
+
+    public readonly string $data;
```
$data should be in AbstractButtonQuery

danog commented on src/EventHandler/ChatButtonQuery.php:
```diff
@@ -0,0 +1,82 @@
+        parent::__construct($API, $rawCallback);
+        $this->chatId = $API->getIdInternal($rawCallback['peer']);
+        $this->messageId = $rawCallback['msg_id'];
+        $this->data = (string) $rawCallback['data'];
```
This should be in AbstractButtonQuery

danog commented on src/EventHandler/AbstractGameQuery.php:
```diff
@@ -0,0 +1,18 @@
+    public function __construct(MTProto $API, array $rawCallback)
+    {
+        parent::__construct($API, $rawCallback);
+        $this->gameShortName = $rawCallback['game_short_name'] ?? '';
```
This should be a not null assertion

danog commented on src/EventHandler/InlineGameQuery.php:
```diff
@@ -0,0 +1,14 @@
+    /** @internal */
+    public function __construct(MTProto $API, array $rawCallback)
+    {
+        parent::__construct($API, $rawCallback);
```
Missing the msg_id

mtalaeii committed:
Small fixes

mtalaeii commented:
Fixed! i hope this is the last change if my code still have bug please say them all in one conversation i will fix that

danog commented on src/EventHandler/AbstractButtonQuery.php:
```diff
@@ -6,8 +6,13 @@
     public function __construct(MTProto $API, array $rawCallback)
     {
         parent::__construct($API, $rawCallback);
+        $this->data = (string) $rawCallback['data'];
```
No need to cast this

mtalaeii replied:
i say my reason and my example in your telegram pv so i close this conversation as resolved

danog commented on src/EventHandler/AbstractQuery.php:
```diff
@@ -0,0 +1,32 @@
+     * @var list<string> Regex matches, if a filter regex is present
+     *
+     */
+    public ?array $matches = null;
```
Move this to AbstractButtonQuery

mtalaeii committed:
Fix
